overlord/devicestate/firstboot.go

```diff
@@ -297,26 +295,29 @@ func populateStateFromSeedImpl(st *state.State, tm timings.Measurer) ([]*state.T
 	return tsAll, nil
 }
 
-func readAsserts(fn string, batch *asserts.Batch) ([]*asserts.Ref, error) {
-	f, err := os.Open(fn)
-	if err != nil {
-		return nil, err
-	}
-	defer f.Close()
-	return batch.AddStream(f)
-}
-
 func importAssertionsFromSeed(st *state.State) (*asserts.Model, error) {
 	// TODO: use some kind of context fo Device/SetDevice?
 	device, err := internal.Device(st)
 	if err != nil {
 		return nil, err
 	}
 
-	// set device,model from the model assertion
 	assertSeedDir := filepath.Join(dirs.SnapSeedDir, "assertions")
-	dc, err := ioutil.ReadDir(assertSeedDir)
-	if release.OnClassic && os.IsNotExist(err) {
+	// collect and
+	// set device,model from the model assertion
+	var modelRef *asserts.Ref
+	checkForModel := func(ref *asserts.Ref) error {
```

Comment:
This is a nice approach! Probably a philosophical question but I was wondering if an alternative approach would be to add something like "Refs()" to asserts.Batch and then iterate those after they are loaded. Its not as efficient as your approach so maybe bad but its (IMHO) slightly easier to read as its more "linear". Something like:
```
        batch, err := seed.LoadAssertions(assertSeedDir, checkForModel)
        if err ....
        ...
        var modelRef *asserts.Ref
	for _, ref := range batch.Refs() {
		if ref.Type == asserts.ModelType {
			if modelRef != nil && modelRef.Unique() != ref.Unique() {
				return nil, fmt.Errorf("cannot add more than one model assertion")
			}
			modelRef = ref
		}
	}
        ....
```
I played a bit with this and http://paste.ubuntu.com/p/GN4ZpY5c7K/ is the full example. Anyway, just an idle wondering - I also don't know how LoadAssertions() will used in addition to this so I may miss important context.

Reply:
as I said it will become in the end an internal helper to seed.